lib/icinga/dependency.hpp

```diff
@@ -42,21 +47,178 @@ class Dependency final : public ObjectImpl<Dependency>
+	 * the order of the composite keys, since they are used to uniquely identify the dependency group. However,
+	 * the order of the keys is not guaranteed when using the actual objects, as they won't point to the same
+	 * address after each reload.
+	 */
+	using CompositeKeyType = std::tuple<String, String, int, bool>;
+
+	/**
+	 * Represents the value type of each dependency group members.
+	 *
+	 * It stores the dependency objects of any given Checkable that produce the same composite key (CompositeKeyType).
+	 * In other words, when looking at the dependency graph from the class description, the two dependency objects
+	 * {C1, C2} -> P1 produce the same composite key, thus they are mapped to the same MemberValueType container with
+	 * "C1" and "C2" as their keys respectively. Since Icinga 2 allows to construct different identical dependencies
+	 * (duplicates), we're using a multimap instead of a simple map here.
+	 */
+	using MemberValueType = std::unordered_multimap<const Checkable*, Dependency*>;
+	using MembersMap = std::map<CompositeKeyType, MemberValueType>;
+
+	DependencyGroup(String name, const Dependency::Ptr& member);
+
+	static DependencyGroup::Ptr Register(const DependencyGroup::Ptr& dependencyGroup);
+	static void Unregister(const DependencyGroup::Ptr& dependencyGroup);
+	static size_t GetRegistrySize();
+
+	static CompositeKeyType MakeCompositeKeyFor(const Dependency::Ptr& dependency);
+
+	/**
+	 * Check whether the current dependency group represents an explicitly configured redundancy group.
+	 *
+	 * @return bool - Returns true if it's a redundancy group, false otherwise.
+	 */
+	inline bool IsRedundancyGroup() const
+	{
+		return !m_Name.IsEmpty();
+	}
+
+	bool HasMembers() const;
+	bool HasIdenticalMember(const Dependency::Ptr& dependency) const;
+	void AddMember(const Dependency::Ptr& member);
+	void RemoveMember(const Dependency::Ptr& member);
+	std::vector<Dependency::Ptr> GetMembers(const Checkable* child) const;
+	size_t GetMemberCount() const;
+
+	void SetIcingaDBIdentifier(const String& identifier);
+	String GetIcingaDBIdentifier() const;
+
+	const String& GetName() const;
+	String GetCompositeKey();
+	String PeekParentCheckableName() const;
+
+	enum State
+	{
+		Unknown           = 1ull << 0,
+		Failed            = 1ull << 1,
+		Unreachable       = 1ull << 2,
+		ReachableOK       = 1ull << 3,
+		UnreachableFailed = Unreachable | Failed,
+	};
+
+	State GetState(DependencyType dt = DependencyState, int rstack = 0) const;
+
+	static boost::signals2::signal<void(const DependencyGroup::Ptr&, const Dependency::Ptr&)> OnMembersChanged;
+
+private:
+	void MoveMembersTo(const DependencyGroup::Ptr& other);
+
+	static size_t Hash(const DependencyGroup::Ptr& dependencyGroup)
+	{
+		return std::hash<String>()(dependencyGroup->GetCompositeKey());
+	}
+
+	static bool Equal(const DependencyGroup::Ptr& lhs, const DependencyGroup::Ptr& rhs)
+	{
+		return lhs->GetCompositeKey() == rhs->GetCompositeKey();
+	}
```

Comment:
If you make these structs with an `operator()` (just like you did in earlier versions with the Boost multi-index containers), this will become nicer to use (you won't need `decltype()` in the type name and also need to pass them to the constructor explicitly).